src/webots/nodes/WbBackground.cpp

```diff
@@ -287,199 +389,248 @@ void WbBackground::applyColourToWren(const WbRgb &color) {
   }
 }
 
-void WbBackground::applySkyBoxToWren() {
-  destroySkyBox();
-
-  WbWrenOpenGlContext::makeWrenCurrent();
-
-  int edgeLength = 0;
-  QString lastFile;
-
-  QString textureUrls[6];
-  QVector<float *> hdrImageData;
-  // 1. Load the background.
-  mCubeMapTexture = wr_texture_cubemap_new();
-  QImage image[6];  // the texture image data should be persistent until the WREN texture cubemap is fully setup
-                    // hence the scope of this QImage array declaration
-  try {
-    bool allUrlDefined = true;
-    bool atLeastOneUrlDefined = false;
-
-    for (int i = 0; i < 6; ++i) {
-      if (mUrlFields[i]->size() == 0) {
-        allUrlDefined = false;
-        textureUrls[i] = "";
-        continue;
-      } else
-        atLeastOneUrlDefined = true;
-
-      textureUrls[i] = WbUrl::computePath(this, "textureBaseName", mUrlFields[i]->item(0), false);
+bool WbBackground::loadTexture(int i) {
+  if (mTexture[i])
+    return true;
+  QString url;
+  QIODevice *device;
+
+  if (mDownloader[i]) {
+    url = mUrlFields[i]->item(0);
+    if (!mDownloader[i]->error().isEmpty()) {
+      warn(tr("Cannot not retrieve '%1': %2").arg(url).arg(mDownloader[i]->error()));
+      delete mDownloader[i];
+      mDownloader[i] = NULL;
+      return false;
     }
+    assert(mDownloader[i]->device());
+    device = mDownloader[i]->device();
+  } else {
+    if (mUrlFields[i]->size() == 0)
+      return false;
+    url = WbUrl::computePath(this, QString("%1Url").arg(gDirections[i]), mUrlFields[i]->item(0), false);
+    if (url == WbUrl::missingTexture()) {
+      warn(tr("Texture not found: '%1'").arg(mUrlFields[i]->item(0)));
+      return false;
+    }
+    device = new QFile(url);
+    if (!device->open(QIODevice::ReadOnly)) {
+      warn(tr("Cannot open texture file: '%1'").arg(url));
+      return false;
```

Comment:
Is it not needed to delete `device` before returning?
Same for the other return cases later in this function.